File: src/visit/AstVisitor.ts
import { AbstractParseTreeVisitor } from 'antlr4ts/tree/AbstractParseTreeVisitor';
import { SqlBaseVisitor } from '../antlr';
import { ColumnReferenceContext } from '../antlr/SqlBaseParser';


export class AstVisitor extends AbstractParseTreeVisitor<object> implements SqlBaseVisitor<object>{
    protected defaultResult(): object {
        return {
            type: 'defaultResult'
        };
    }
    // 自底向上，从unquotedIdentifier开始
    visitUnquotedIdentifier(ctx: any) {
        return {
            type: 'Identifier',
            name: ctx.text
        }
    }
    // querySpecification
    visitQuerySpecification(ctx: any) {
        // 分出selectItem组成的数组和relation组成的数组
        const selectItems = ctx.selectItem();
        const relations = ctx.relation();
        // 从selectItem中获取identifier
        const selectList = selectItems.map((selectItem: any) => {
            return this.visit(selectItem);
        });
        // √ console.log('selectList', selectList)
        // 从relation中获取identifier
        const fromClause = relations.map((relation: any) => {
            return this.visit(relation);
        });
        // √ console.log('fromClause', fromClause)
        return {
            type: 'QuerySpecification',
            queryFileds: selectList,
            tables: fromClause
        }
    }
    visitSingleStatement(ctx: any) {
        // console.log('single-statement', ctx.text)
        // console.table(ctx.children)
        const obj = super.visit(ctx.children[0]); // important
        // console.log('single-statement', obj)
        return obj;
    }
    visitStatement(ctx: any) {
        return super.visitChildren(ctx);
    }
    visitStatementDefault(ctx: any) {
        // console.log('statement-default', ctx.text)
        const obj = super.visitChildren(ctx);
        // console.log('statement-default', obj)
        return obj;
    }
    visitQuery(ctx: any) {
        const obj = super.visitChildren(ctx);
        //√ console.log('query', obj)
        return obj;
    }
    visitQueryNoWith(ctx: any) {
        const obj = super.visitChildren(ctx);
        // console.log('query-no-with', obj)
        return obj;
    }
    visitQueryTerm(ctx: any) {
        return super.visitChildren(ctx);
    }
    visitQueryTermDefault(ctx: any) {
        return super.visitChildren(ctx);
    }
    visitQueryPrimary(ctx: any) {
        return super.visitChildren(ctx);
    }
    visitQueryPrimaryDefault(ctx: any) {
        const obj = super.visitChildren(ctx);
        // √ console.log('query-primary-default', obj)
        return obj;
    }
    visitSelectItem(ctx: any) {
        return super.visitChildren(ctx);
    }
    visitSelectSingle(ctx: any) {
        return super.visitChildren(ctx);
    }
    visitExpression(ctx: any) {
        return super.visitChildren(ctx);
    }
    visitBooleanExpression(ctx: any) {
        return super.visitChildren(ctx);
    }
    visitPredicated(ctx: any) {
        return super.visitChildren(ctx);
    }
    visitValueExpression(ctx: any) {
        return super.visitChildren(ctx);
    }
    visitValueExpressionDefault(ctx: any) {
        const obj = super.visitChildren(ctx);
        // console.log('value-expression-default', obj)
        // console.table(obj)
        return obj;
    }
    visitPrimaryExpression(ctx: any) {
        return super.visitChildren(ctx);
    }
    visitColumnReference(ctx: any) {
        const obj = super.visitChildren(ctx);
        // console.log('column-reference children', obj)
        // console.table(obj)
        return obj;
    }
    visitIdentifier(ctx: any) {
        return super.visitChildren(ctx);
    }
    visitRelation(ctx: any) {
        return super.visitChildren(ctx);
    }
    visitRelationDefault(ctx: any) {
        return super.visitChildren(ctx);
    }
    visitSampledRelation(ctx: any) {
        return super.visitChildren(ctx);
    }
    visitAliasedRelation(ctx: any) {
        return super.visitChildren(ctx);
    }
    visitRelationPrimary(ctx: any) {
        return super.visitChildren(ctx);
    }
    visitTableName(ctx: any) {
        return super.visitChildren(ctx);
    }
    visitQualifiedName(ctx: any) {
        return super.visitChildren(ctx);
    }
}

File: demo/src/main/java/com/example/visit/AstVisitor.java
package com.example.visit;

import java.util.List;
import static java.util.stream.Collectors.toList;

import org.antlr.v4.runtime.ParserRuleContext;

import com.example.antlr4.SqlBaseBaseVisitor;
import com.example.antlr4.SqlBaseParser;
import com.example.antlr4.SqlBaseParser.RelationContext;
import com.example.antlr4.SqlBaseParser.SelectItemContext;
import com.example.tree.basic.Node;
import com.example.tree.detail.Identifier;
import com.example.tree.detail.QuerySpecification;

public class AstVisitor extends SqlBaseBaseVisitor<Node> {
    @Override
    public Node visitSingleStatement(SqlBaseParser.SingleStatementContext ctx) {
        return visit(ctx.statement());
    }

    @Override
    public Node visitUnquotedIdentifier(SqlBaseParser.UnquotedIdentifierContext ctx) {
        return new Identifier(ctx.getText(), false);
    }

    @Override
    public Node visitQuerySpecification(SqlBaseParser.QuerySpecificationContext ctx) {
        // ctx.children 分为 selectItem组 和 relation组
        List<SelectItemContext> selectItemsCtx = ctx.selectItem();
        List<RelationContext> relationCtx = ctx.relation();
        // 1. 处理 selectItem组
        List<Identifier> selectItems = visitChildren(selectItemsCtx, Identifier.class);
        // 2. 处理 relation组
        List<Identifier> relations = visitChildren(relationCtx, Identifier.class);
        // 3. 组合成一个Node
        QuerySpecification querySpecification = new QuerySpecification(selectItems, relations);
        return querySpecification;
    }

    // 自己实现一个visit方法，用于遍历节点的同类子节点
    private <T> List<T> visitChildren(List<? extends ParserRuleContext> ctxs, Class<T> clazz) {
        return ctxs.stream().map(this::visit).filter(clazz::isInstance).map(clazz::cast).collect(toList());
    }
}
